File: algorithms/php/74.Searcha2DMatrix.php
<?php

function searchMatrix($matrix, $target)
{
    $rows = count($matrix);
    $cols = count($matrix[0]);

    $l = 0;
    $r = $cols * $rows - 1;

    while ($l <= $r) {
        $m = intval(($l + $r) / 2);


        $mc = $m % $cols;
        $mr = intval($m / $cols);

        if ($target === $matrix[$mr][$mc]) {
            return true;
        }

        if ($target < $matrix[$mr][$mc]) {
            $r = $m - 1;
        }

        if ($target > $matrix[$mr][$mc]) {
            $l = $m + 1;
        }
    }

    return false;
}

File: algorithms/cpp/74.Searcha2DMatrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        int l = 0;
        int r = cols * rows - 1;

        while (l <= r) {
            int m = l + (r - l) / 2;

            int mc = m % cols;
            int mr = m / cols;

            if (matrix[mr][mc] == target) {
                return true;
            }

            if (matrix[mr][mc] < target) {
                l = m + 1;
            }

            if (matrix[mr][mc] > target) {
                r = m - 1;
            }
        }

        return false;
    }
};
